docs/plugins.md

```diff
@@ -184,20 +236,162 @@ function embed(
+Comments are then attached to these nodes **mutating the AST in the
+process** and the `comments` property is deleted from the AST root.
+The `*Comment` functions are used to adjust Prettier's
+algorithm. Once attached to the AST, Prettier will automatically call
+the `printComment(path, options): Doc` function and insert the returned
+doc into the (hopefully) correct place.
+
+#### (optional) `printComment`
+
+Called whenever a comment node needs to be printed. It has the signature:
 
 ```ts
-function preprocess(ast: AST, options: object): AST;
+function printComment(
+  // Path to the current comment node
+  commentPath: FastPath,
+  // Print a node with the current printer
+  options: object
+): Doc;
+```
+
+#### (optional) `canAttachComment`
+
+```ts
+function canAttachComment(node: AST): boolean;
 ```
 
+This function is used for deciding whether a comment can be attached to
+a particular AST node. By default, _all_ AST properties are traversed
+searching for nodes that comments can be attached to. This function
+is used to prevent comments from being attached to a particular node.
+A typical implementation looks like
+
+```js
+function canAttachComment(node) {
+  return node.type && node.type !== "comment";
+}
+```
+
+#### (optional) `isBlockComment`
+
+```ts
+function canAttachComment(node: AST): boolean;
+```
+
+Returns whether or not the AST node is a block comment.
+
+#### (optional) `handleComments`
+
+The `handleComments` object contains three optional functions, each with signature
+
+```ts
+function(
+	// The AST node corresponding to the comment
+	comment: AST,
+	// The full source code text
+	text: string,
+	// The global options object
+	options: object,
+	// The AST
+	ast: AST,
+	// Whether this comment is the last comment
+	isLastComment: boolean
+): boolean
+```
+
+These functions are used to override Prettier's default comment
+assignment algorithm. `ownLine`/`endOfLine`/`remaining` is expected
+to either manually assign a comment to a node and return `true`,
+or return a `falsy` value and let Prettier assign the comment.
+
+Based on the text surrounding a comment node, Prettier dispatches:
+
+    * `ownLine` if a comment has only whitespace preceding it and a newline afterwards,
+    * `endOfLine` if a comment has a newline afterwards but some non-whitespace preceding it,
+    * `remaining` in all other cases.
+
+At the time of dispatching, Prettier will have annotated each AST comment
+node (i.e., created new properties) with at least one of `enclosingNode`,
+`precedingNode`, or `followingNode`. These can be used to aid a plugin's
+decision process (of course the entire AST and original text is also
+passed in for making more complicated decisions).
+
+_Manually assigning a comment_
+
+Nodes with comments are expected to have a `comments` property containing
+an array of comments. Each comment is expected to have the following
+properties: `leading`, `trailing`, `printed`. An example `ownLine`
```

Comment:
These details need to be de-emphasized in favor of `util.add*Comment` as they can change because of https://github.com/prettier/prettier/issues/8122 whereas `util.add*Comment` will stay.